firmware/defmt-rtt/src/lib.rs

```diff
@@ -39,13 +37,12 @@ struct Logger;
 
 /// Global logger lock.
 static TAKEN: AtomicBool = AtomicBool::new(false);
-static INTERRUPTS_ACTIVE: AtomicBool = AtomicBool::new(false);
+static INTERRUPTS_ACTIVE: AtomicU8 = AtomicU8::new(0);
 static mut ENCODER: defmt::Encoder = defmt::Encoder::new();
 
 unsafe impl defmt::Logger for Logger {
     fn acquire() {
-        let primask = register::primask::read();
-        interrupt::disable();
+        let token = unsafe { critical_section::acquire() };
```

Comment:
@jannic Could you shed light on this?

Reply:
As @jonas-schievink already mentioned, `critical-section` is missing some documentation.
This is part of it: The intended safety-guarantees are not specified.

As far as I can tell, there are no real requirements. The current implementations of `critical_section::acquire()` are notreally unsafe, but just disable interrupts.

My guess is that it's just for symmetry with `critical_section::release()`, which is obviously unsafe, as enabling interrupts could break other code expecting to run exclusively.

@Dirbaio, any comments?

Reply:
@jannic Could you please add safety comments for `critical_section::acquire` and `critical_section::release`, with what you wrote here? Then we should be good to go.

Reply:
I still have to write docs for `critical-section` (sorry! :see_no_evil: ), but the safety contract is essentially:

- Each acquire must be paired with a release with the same token.
- acquire/release pairs must be "properly nested", ie it's not OK to do `a=acquire(); b=acquire(); release(a); release(b);`.

Code in this PR complies with the safety contract, so LGTM :+1:

Reply:
`acquire()` has to be unsafe because you can use it to break the "properly nested" requirement in *someone else's* acquire/release pair deeper in the call stack, if you don't do the corresponding `release()` call.

Reply:
Thanks, that sounds good for now.

Reply:
> As `fn acquire()` is not unsafe, one could break this contract by just calling `Logger::acquire()` in some random location. 

One can't, as `Logger::acquire()` is private.
I still think that it wouldn't do harm if one could call it, but as one can't, either way it doesn't matter.

Reply:
Oops, too fast...
There is `defmt::export::acquire();`, which is `pub` and not `unsafe`. So it's perfectly possible to call `acquire()` out of order. AFAIK the only thing bad which will happen is that the next call to a defmt logger will panic ("defmt logger taken reentrantly"), which seems to be a perfectly fine response.

However, there is something much more dangerous: `defmt::export::release()`. which allows to end a critical section unconditionally (or, before merging this pull request, to enable interrupts). I think this should be `unsafe`, right?
```
info!("something");  // log something while interrupts are enabled. `INTERRUPTS_ACTIVE` remembers that interrupts were enabled
cortex_m::interrupt::free(|cs| {
  // interrupts are disabled
  defmt::export::release();
  // oh no! interrupts are enabled, again
  // but we still have the cs token:
  let borrowed = cortex_m_mutex.borrow(cs);
  // do something with shared value while interrupts are enabled
}
```